proposals/0005-fallback-rounting.md

```diff
@@ -0,0 +1,229 @@
+an intrinsic feature of Glide routing despite routing strategy.
+
+There may be exceptions to this dictated by the model nature.
+For example, embedding API may require one model to be used for all embeddings produced.
+
+Based on this, we group all models we support into types (see GEP0002):
+
+- language models
+- embeddings
+- speech transcribers
+- speech synthesizers
+- etc.
+
+Each model type is represented as a list of model pools that could serve the incoming request uniformly.
+Each model pool is a list of models (along with their params & operational info like API key) and routing strategy that defines
+how incoming requests should be served.
+
+The model pool ID or name is exposed to clients and they provide IDs during requests (see GEP0002).
+This is how Glide understands which one should serve the concrete request.
+
+### Model Routing & Fallbacking
+
+Glide leverages fallback models on the first fail and do adaptive health tracking (via token bucket). 
+Moving right to the next model serves as a retry for the first one while we spend as little time as possible waiting. 
+
+Then, the whole service instance is going to be aware about number of fails that happen to a certain model 
+and as soon as that number goes over the roof, 
+the model is going to be "disqualified" to serve requests for some time. 
+This means that as soon as the model is dead we pull it out of the pool, so the next requests are going to be served by healthy models only. 
+This strategy should lead to much better latency on failover.
+
+![Routing & Fallbacks](./imgs/glide-fallback-logic.png)
+
+### Routing Strategies
+
+#### Priority
+
+Route tracking to the first healthy model in the order in which the user defined models in the router setup.
+So if the first model is healthy it's going to handle all requests until it's unavailable. 
+Then, the second model will handle all requests, and so on.
+
+If the first model is available again, all traffic will go to it.
+
+#### Round Robin
+
+Iterate through model list in cycle. 
+If the model is not available, it will iterate again until it finds a healthy model or make a full cycle with no luck.
+
+### Weighted Round Robin
+
+Send each model traffic proportional to the model weight.
+
+If a model become unavailable, it's taken out of consideration completely until it's available again.
+So the traffic split was:
+- model A: 8
+- model B: 1
+- model C 1
+
+And the model A becomes unavailable, then model B & C will handle an equal portion of all traffic.
+
+### Least Latency
+
+Picks a healthy model with the least average latency over time. 
+If the least latency model becomes unhealthy, it will pick the second the best, etc.
+
+Since we don't know the true distribution of model latencies,
+and it's likely a moving value, we do some attempts to estimate it and keep updated over time.
+
+The algorithm consists of two stages:
```

Comment:
So the latencies are determined based on the same request being sent to multiple models? I noted this elsewhere but the response latency is in part due to the quantity of tokens generated. More tokens generated means slower response time (generally speaking). We might want to incorporate something to account for tokens generated and it's effect on latency.

Reply:
@mkrueger12 yes, this is still an open concern. I have filled a follow-up ticket to do that normalization: https://github.com/EinStack/glide/issues/78